File: .github/scripts/fetch_library.py
#!/usr/bin/env python3
"""获取 Steam 游戏库数据：GetOwnedGames + 批量 appdetails → library.json"""

import os
import json
from pathlib import Path
from common import filter_library_games, get_steam_id, get_owned_games, fetch_steam_details, fetch_review, batch_fetch, build_games_output

STEAM_API_KEY = os.environ.get('STEAM_API_KEY', '')
STEAM_USER_ID = os.environ.get('STEAM_USER_ID', '')
STEAM_LANG = os.environ.get('STEAM_LANG', 'schinese')


def main():
    base_dir = Path(__file__).parent.parent.parent

    print('=' * 60)
    print('Steam 游戏库数据获取')
    print('=' * 60)

    print('\n1. 获取 Steam ID...')
    steam_id = get_steam_id(STEAM_API_KEY, STEAM_USER_ID)
    if not steam_id:
        print('   获取 Steam ID 失败')
        return
    print(f'   Steam ID: {steam_id}')

    print('\n2. 获取游戏库...')
    owned, total_count = get_owned_games(STEAM_API_KEY, steam_id)
    if not owned:
        print('   游戏库为空')
        return
    print(f'   共 {total_count} 款游戏')

    print(f'\n3. 获取游戏详情 ({len(owned)} 款)...')
    appids = [g['appid'] for g in owned]
    playtime_map = {g['appid']: g['playtime_hours'] for g in owned}

    detail_map = batch_fetch(appids, lambda aid: fetch_steam_details(aid, STEAM_LANG), progress_interval=20)
    print(f'\n4. 获取评测数据 ({len(detail_map)} 款)...')
    review_appids = sorted(detail_map.keys(), key=lambda a: playtime_map.get(a, 0), reverse=True)
    review_map = batch_fetch(review_appids, lambda aid: fetch_review(aid, STEAM_LANG))

    print(f'   已获取 {len(review_map)} 款游戏的评测')

    print('\n5. 合并数据...')
    library_games = []
    for g in owned:
        aid = g['appid']
        detail = detail_map.get(aid, {})
        library_games.append({
            'appid': aid,
            'name': detail.get('name', g['name']),
            'playtime_hours': g['playtime_hours'],
            'header_image': detail.get('header_image', ''),
            'short_description': detail.get('short_description', ''),
            'genres': detail.get('genres', []),
            'screenshots': detail.get('screenshots', []),
            'review': review_map.get(aid),
        })

    total_playtime = sum(g['playtime_hours'] for g in library_games)
    library_games, software_count, filtered_count, total_playtime = filter_library_games(library_games, detail_map)
    if software_count > 0:
        print(f'   过滤掉 {software_count} 款非游戏（软件/视频等）')
    if filtered_count > 0:
        print(f'   过滤掉 {filtered_count} 款游玩时长不足总时长千分之一的游戏')

    output = build_games_output(library_games)

    library_file = base_dir / 'library.json'
    with open(library_file, 'w', encoding='utf-8') as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f'\n✓ library.json 已生成 ({len(library_games)} 款游戏, {total_playtime:.1f} 小时)')


if __name__ == '__main__':
    main()

File: worker/scripts/fetch-library.ts
import { KV_KEYS, getSteamId, getOwnedGames, fetchSteamDetails, fetchReview, batchFetch, filterLibraryGames, buildGamesOutput, getConfig } from '../lib/steam.js'

export async function fetchLibrary(env: Env): Promise<void> {
  const steamApiKey = await getConfig(env, 'STEAM_API_KEY')
  const steamUserId = await getConfig(env, 'STEAM_USER_ID')
  const lang = await getConfig(env, 'STEAM_LANG', 'schinese')

  console.log('获取 Steam ID...')
  const steamId = await getSteamId(steamApiKey, steamUserId)
  if (!steamId) throw new Error('获取 Steam ID 失败')

  console.log('获取游戏库...')
  const { games: owned, count: totalCount } = await getOwnedGames(steamApiKey, steamId)
  if (!owned.length) throw new Error('游戏库为空')
  console.log(`共 ${totalCount} 款游戏`)

  console.log(`获取游戏详情 (${owned.length} 款)...`)
  const appids = owned.map(g => g.appid)
  const playtimeMap: Record<number, number> = {}
  for (const g of owned) playtimeMap[g.appid] = g.playtime_hours

  const detailMapRaw = await batchFetch(appids, aid => fetchSteamDetails(aid, lang), { maxWorkers: 20, delay: 0.2 })
  const detailMap = detailMapRaw as Record<number, Record<string, unknown>>

  console.log(`获取评测数据 (${Object.keys(detailMap).length} 款)...`)
  const reviewAppids = Object.keys(detailMap).map(Number)
    .sort((a, b) => (playtimeMap[b] || 0) - (playtimeMap[a] || 0))
    .slice(0, 50)
  const reviewMap = await batchFetch(reviewAppids, aid => fetchReview(aid, lang), { maxWorkers: 10, delay: 0.2 })

  console.log('合并数据...')
  const libraryGames = owned.map(g => ({
    appid: g.appid,
    name: (detailMap[g.appid]?.name as string) || g.name,
    playtime_hours: g.playtime_hours,
    header_image: (detailMap[g.appid]?.header_image as string) || '',
    short_description: (detailMap[g.appid]?.short_description as string) || '',
    genres: (detailMap[g.appid]?.genres as string[]) || [],
    screenshots: (detailMap[g.appid]?.screenshots as string[]) || [],
    review: (reviewMap[g.appid] as Record<string, unknown>) || null,
  }))

  const detailTypeMap: Record<number, { type?: string }> = {}
  for (const aid of Object.keys(detailMap).map(Number)) {
    detailTypeMap[aid] = { type: detailMap[aid]?.type as string | undefined }
  }

  const { games: filteredGames, softwareCount, filteredCount } = filterLibraryGames(libraryGames, detailTypeMap)
  if (softwareCount > 0) console.log(`过滤掉 ${softwareCount} 款非游戏`)
  if (filteredCount > 0) console.log(`过滤掉 ${filteredCount} 款低时长游戏`)

  const output = buildGamesOutput(filteredGames)
  await env.KV.put(KV_KEYS.DATA_LIBRARY, JSON.stringify(output))
  const totalPlaytime = filteredGames.reduce((s, g) => s + (g.playtime_hours || 0), 0)
  console.log(`✓ library.json 已生成 (${filteredGames.length} 款游戏, ${totalPlaytime.toFixed(1)} 小时)`)
}
